FirebaseAppCheck/Sources/AppAttestProvider/Storage/FIRAppAttestArtifactStorage.m

```diff
@@ -22,16 +22,82 @@
 #import "FBLPromises.h"
 #endif
 
+#import <GoogleUtilities/GULKeychainStorage.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+static NSString *const kKeychainService = @"com.firebase.app_check.app_attest_artifact_storage";
+
+@interface FIRAppAttestArtifactStorage ()
+
+@property(nonatomic, readonly) NSString *appName;
+@property(nonatomic, readonly) NSString *appID;
+@property(nonatomic, readonly) GULKeychainStorage *keychainStorage;
+@property(nonatomic, readonly, nullable) NSString *accessGroup;
+
+@end
+
 @implementation FIRAppAttestArtifactStorage
 
-- (nonnull FBLPromise<NSData *> *)getArtifact {
-  // TODO: Implement.
-  return [FBLPromise resolvedWith:nil];
+- (instancetype)initWithAppName:(NSString *)appName
+                          appID:(NSString *)appID
+                keychainStorage:(GULKeychainStorage *)keychainStorage
+                    accessGroup:(nullable NSString *)accessGroup {
+  self = [super init];
+  if (self) {
+    _appName = [appName copy];
+    _appID = [appID copy];
+    _keychainStorage = keychainStorage;
+    _accessGroup = [accessGroup copy];
+  }
+  return self;
 }
 
-- (nonnull FBLPromise<NSData *> *)setArtifact:(nonnull NSData *)artifact {
-  // TODO: Implement.
-  return [FBLPromise resolvedWith:nil];
+- (instancetype)initWithAppName:(NSString *)appName
+                          appID:(NSString *)appID
+                    accessGroup:(nullable NSString *)accessGroup {
+  GULKeychainStorage *keychainStorage =
+      [[GULKeychainStorage alloc] initWithService:kKeychainService];
+  return [self initWithAppName:appName
+                         appID:appID
+               keychainStorage:keychainStorage
+                   accessGroup:accessGroup];
+}
+
+- (FBLPromise<NSData *> *)getArtifact {
+  return [self.keychainStorage getObjectForKey:[self artifactKey]
+                                   objectClass:[NSData class]
+                                   accessGroup:self.accessGroup]
+      .then(^NSData *(id<NSSecureCoding> storedArtifact) {
+        if ([(NSObject *)storedArtifact isKindOfClass:[NSData class]]) {
+          return (NSData *)storedArtifact;
+        } else {
+          return nil;
```

Comment:
Should a promised rejected with some error be returned instead?

Reply:
Maybe... The method should return `nil` if the artifact was not found. The case when something else was stored in the keychain looks more similar to "not found" case so we eventually should overwrite the value. WDYT?